NetworkParameters: Change PacketMagic to an `int`

msgilligan opened the pull request:
It's just a binary value and does not have signed-edness.

Replace `writeUint32BE` methods in `ByteUtils` with `writeInt32BE` methods

schildbach commented:
Will work on this on Tuesday.

schildbach commented:
`BitcoinSerializer` and `BlockFileLoader`  do some crazy bitshifting with the packet magic. I'm not sure if that stuff still works on int.

Just an idea, but perhaps we can simplify those usecases first?

schildbach commented:
In BlockFileLoader, we could perhaps read an entire 32-bit integer in the right order and then compare it to the packet magic in one comparison?

msgilligan commented:
> `BitcoinSerializer` and `BlockFileLoader` do some crazy bitshifting with the packet magic. I'm not sure if that stuff still works on int.

They are all unsigned shifts (`>>>`) of 24-bits or left and they are all masking with `0xff` to extract the low-order byte, so they work the same on `int` as on `long`.

All unit tests are passing.

> Just an idea, but perhaps we can simplify those usecases first?

What do you have in mind for simplification?  (I've got one related PR coming...)

msgilligan commented:
> In BlockFileLoader, we could perhaps read an entire 32-bit integer in the right order and then compare it to the packet magic in one comparison?

We can't do that because the packetMagic is matched on a byte-by-byte basis. (I suppose we could read a 32-bit value compare and then reverse the buffer pointer by 3 bytes, but since we're matching byte-by-byte we should probably read byte-by-byte)

schildbach commented:
Can you rebase this on current master? For me it fails a number of tests.

schildbach commented:
Other than that, I'm fine to merge.

msgilligan commented:
> Can you rebase this on current master? For me it fails a number of tests.

I rebased and saw the test failures. `packetMagic` was the last use case of writeInt32BE(long, ...) and was failing the sign
check in those methods. So I replaced them with writeInt32BE(int, ...) equivalents.

It's now passing all tests and has been force-pushed.

schildbach commented on core/src/main/java/org/bitcoinj/base/internal/ByteUtils.java:
```diff
@@ -221,34 +221,26 @@ public static void writeInt32LE(long val, byte[] out, int offset) throws ArrayIn
      *                                        if the value doesn't fit the remaining buffer
      */
-    public static void writeInt32BE(long val, byte[] out, int offset) throws ArrayIndexOutOfBoundsException {
-        check(offset >= 0 && offset <= out.length - 4, () ->
```
Please don't remove these checks and comments.

msgilligan replied:
ded back and force-pushed.

schildbach commented on core/src/main/java/org/bitcoinj/base/internal/ByteUtils.java:
```diff
@@ -346,14 +338,12 @@ public static void writeInt32LE(long val, OutputStream stream) throws IOExceptio
 
     /**
      * Write a 32-bit integer to a given output stream in big-endian format.
-     * <p>
```
Please don't remove this comment.

msgilligan replied:
Added back and force-pushed.

msgilligan commented:
> I think we have methods for writing signed ints already.

Not for the `writeInt32BE(int...)` case. I have updated the PR by adding those methods and leaving the `long` methods in place.

msgilligan commented on core/src/main/java/org/bitcoinj/base/internal/ByteUtils.java:
```diff
@@ -221,34 +221,26 @@ public static void writeInt32LE(long val, byte[] out, int offset) throws ArrayIn
      * @throws BufferOverflowException if the value doesn't fit the remaining buffer
      */
-    public static ByteBuffer writeInt32BE(long val, ByteBuffer buf) throws BufferOverflowException {
-        checkArgument(val >= 0 && val <= MAX_UNSIGNED_INTEGER, () ->
```
Added back and force-pushed.

schildbach commented:
Merged.
